base/strings/search.jl

```diff
@@ -23,176 +44,190 @@ julia> search("Hello to the world", "z")
 
 julia> search("JuliaLang","Julia")
 1:5
+
+julia> search(Int8[1,2,3], Int8[2,3])
+2:3
+
+julia> search("Hello to the world", 'z')
+0
+
+julia> search("Hello to the world", ['e', 'l'])
+2
+
+julia> search("", ['e', 'l'])
+0
 ```
 """
 function search(s::AbstractString, c::Chars, i::Integer)
-    if isempty(c)
-        return 1 <= i <= nextind(s,endof(s)) ? i :
-               throw(BoundsError(s, i))
-    end
-    if i < 1 || i > nextind(s,endof(s))
-        throw(BoundsError(s, i))
-    end
-    while !done(s,i)
-        d, j = next(s,i)
+    isvalid(s, i) || throw(ArgumentError("index $i is invalid"))
+    isempty(c) && return 0
+    while !done(s, i)
+        d, j = next(s, i)
         if d in c
             return i
         end
         i = j
     end
     return 0
 end
-search(s::AbstractString, c::Chars) = search(s,c,start(s))
 
-in(c::Char, s::AbstractString) = (search(s,c)!=0)
+search(s::AbstractString, c::Chars) = s == "" ? 0 : search(s, c, start(s))
+
+in(c::Char, s::AbstractString) = search(s,c) != 0
 
 function _searchindex(s, t, i)
-    if isempty(t)
-        return 1 <= i <= nextind(s,endof(s)) ? i :
-               throw(BoundsError(s, i))
-    end
+    # assumes that caller has checked if i is a valid index
+    isempty(t) && return i
+
     t1, j2 = next(t,start(t))
     while true
-        i = search(s,t1,i)
-        if i == 0 return 0 end
-        c, ii = next(s,i)
+        i = search(s, t1, i)
+        i == 0 && return 0
+        c, ii = next(s, i)
         j = j2; k = ii
         matched = true
-        while !done(t,j)
-            if done(s,k)
+        while !done(t, j)
+            if done(s, k)
                 matched = false
                 break
             end
-            c, k = next(s,k)
-            d, j = next(t,j)
+            c, k = next(s, k)
+            d, j = next(t, j)
             if c != d
                 matched = false
                 break
             end
         end
-        if matched
-            return i
-        end
+        matched && return i
         i = ii
     end
 end
 
-_searchindex(s, t::Char, i) = search(s, t, i)
-
 function _search_bloom_mask(c)
     UInt64(1) << (c & 63)
 end
 
-_nthbyte(s::String, i) = codeunit(s, i)
-_nthbyte(a::ByteArray, i) = a[i]
+_nthbyte(s::String, i) = @inbounds codeunit(s, i)
```

Comment:
Better add `@inbounds` at the call site, and add `@propagate_inbounds` annotations here.